infrastructure/src/main/java/org/corfudb/infrastructure/AutoCommitService.java

```diff
@@ -0,0 +1,217 @@
+                        getCorfuRuntime().getAddressSpaceView().prefixTrim(trimToken, false);
+                        committedTail = trimMark.getSequence() - 1;
+                    }
+                    lastEpoch = currEpoch;
+                    trimmed = false;
+                }
+
+                log.debug("runAutoCommit: trying to commit [{}, {}].", committedTail + 1, lastLogTail);
+
+                // Commit addresses in batches, retry limit is shared by all batches in this cycle.
+                // NOTE: This implementation relies on the fact that state transfer invokes the
+                // read protocol and fill holes, otherwise the committedTail could be invalid.
+                // (e.g. 1. State transferred a hole at address 100 from A to B; 2. Commit address
+                // 100, which only goes to A as B is not in this address segment; 3. B finishes
+                // state transfer and segments merged; 4. Send a new committedTail to B which is
+                // invalid as 100 is still a hole on B.
+                while (committedTail < lastLogTail) {
+                    long commitStart = committedTail + 1;
+                    long commitEnd = Math.min(committedTail + COMMIT_BATCH_SIZE, lastLogTail);
+                    getCorfuRuntime().getAddressSpaceView().commit(commitStart, commitEnd);
+                    log.trace("runAutoCommit: successfully committed batch [{}, {}]", committedTail, commitEnd);
+                    committedTail = commitEnd;
+                }
+
+                log.debug("runAutoCommit: successfully finished auto commit cycle. " +
+                        "New committed tail: {}.", committedTail);
+                break;
+
+            } catch (RuntimeException re) {
+                log.warn("runAutoCommit: encountered an exception on attempt {}/{}.",
+                        i, MAX_COMMIT_RETRY, re);
+
+                if (i >= MAX_COMMIT_RETRY) {
+                    log.warn("runAutoCommit: retry exhausted, abort and wait for next cycle");
+                    break;
+                }
+
+                // When inspecting address, log unit server could throw a TrimmedException if
+                // the address to inspect is trimmed, then we need to adjust committed tail.
+                if (re instanceof TrimmedException) {
+                    trimmed = true;
+                }
+
+                if (re instanceof NetworkException || re.getCause() instanceof TimeoutException) {
+                    Sleep.sleepUninterruptibly(CONN_RETRY_RATE);
+                }
+
+            } catch (Throwable t) {
+                log.error("runAutoCommit: encountered unexpected exception", t);
+                updateLastLogTail(currentLayout);
+                throw t;
+            }
+        }
+
+        // Update lastLogTail no matter commit cycle succeeds or not.
+        updateLastLogTail(currentLayout);
+    }
+
+    private Layout updateLayoutAndGet() {
```

Comment:
what's the motivation, why you need to update the layout?
FD is in charge of doing it, not sure if it safe to update the layout here

Reply:
Layout need to be updated to handle wong epoch, and disable auto commit if not primary sequencer. Safety is guarantee by using layoutHelper. I removed saveManagementLayout as it's not important.